python/cudf/cudf/core/frame.py

```diff
@@ -2069,7 +2069,12 @@ def from_arrow(cls, data):
                 else:
                     dtype = np_dtypes[name]
 
-                result._data[name] = result._data[name].astype(dtype)
+                if cudf.utils.dtypes.is_struct_dtype(dtype):
+                    result._data[name] = result._data[name]._rename_fields(
+                        [field.name for field in data[name].type]
+                    )
+                elif not cudf.utils.dtypes.is_list_dtype(dtype):
+                    result._data[name] = result._data[name].astype(dtype)
```

Comment:
This can typecast through a copy which would be expensive. What case is this supposed to handle?

Reply:
This block of code was introduced to specifically handle some non-pandas dtypes where arrow would encode the dtypes differently for `categorical`, `boolean`, `object` & `struct`. Previously the type-cast was happening for all dtypes. Now I have changed these type-casts to happen for only when there is this special case handling needed for above-mentioned types.

Reply:
I'm not following the logic here. Are these zero copy typecasts or are we actually changing the memory? Can you give a couple examples of situations when this typecasting would happen?

Reply:
Explained the special cases below:

```python
            for name in result._data.names:
                dtype = None
                if len(result._data[name]) == 0 and pandas_dtypes[name] == "categorical":
                    # CASE 1: Empty Categorical Column
                    dtype = "category"
                elif (
                    pandas_dtypes[name] == "empty"
                    and np_dtypes[name] == "object"
                ):
                    # CASE 2: Column with all nulls and dtype is "object"
                    dtype = np_dtypes[name]
                elif pandas_dtypes[
                    name
                ] == "object" and cudf.utils.dtypes.is_struct_dtype(
                    np_dtypes[name]
                ):
                    # Case 3: Column is a Struct Column
                    result._data[name] = result._data[name]._rename_fields(
                        [field.name for field in data[name].type]
                    )
                else:
                    # No special handling means the dtype already
                    # exists in cudf, hence there is no need to type-cast.
                    continue
```

Case 1:
When there is a categorical column in pyarrow table with length 0, our `from_arrow` libcudf call is returning a column with `int8` dtype(Possibly codes column). To bypass this issue case 1 is inplace. cc: @rgsl888prabhu 

Case 2: 
When all values in a column are `null`, pyarrow encodes a string column as `empty` instead of true dtype, but then the true dtype is stored in `np_dtypes`:
```python
>>> import cudf
>>> df = cudf.DataFrame({'a':cudf.Series([None,  None, None], dtype='str')})
>>> df['j'] = cudf.Series([None, None, None], dtype='uint64')
>>> df
      a     j
0  <NA>  <NA>
1  <NA>  <NA>
2  <NA>  <NA>
>>> table = df.to_arrow()
>>> table.schema.pandas_metadata
{'index_columns': [{'kind': 'range', 'name': None, 'start': 0, 'stop': 3, 'step': 1}], 'column_indexes': [{'name': None, 'field_name': None, 'pandas_type': 'unicode', 'numpy_type': 'object', 'metadata': {'encoding': 'UTF-8'}}], 'columns': [{'name': 'a', 'field_name': 'a', 'pandas_type': 'empty', 'numpy_type': 'object', 'metadata': None}, {'name': 'j', 'field_name': 'j', 'pandas_type': 'uint64', 'numpy_type': 'uint64', 'metadata': None}], 'creator': {'library': 'pyarrow', 'version': '1.0.1'}, 'pandas_version': '1.1.5'}
>>> df.dtypes
a    object
j    uint64
dtype: object
>>> cudf.DataFrame.from_arrow(table).dtypes
a      int8
j    uint64
dtype: object
```
To avert this issue, Case 2 has been added.

Case 3: 
We will need to rename the struct fields as libcudf will not be aware of the field names. 

The `boolean` issue seems to be handled by libcudf layer now so removed special handling at python level.

Reply:
Can you add comments to this effect inline in the code? I think this is good to go afterwards. Thanks for tackling this!

Reply:
Done, Updated the inline comments 👍